opencog/atoms/distvalue/CTHist.cc

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (C) 2019 SingularityNet
+ * All Rights Reserved
+ *
+ * Written by Roman Treutlein
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <algorithm>
+#include <cmath>
+
+#include <opencog/util/exceptions.h>
+
+#include <opencog/atoms/distvalue/CTHist.h>
+
+using namespace opencog;
+
+template <typename val_t>
+void CTHist<val_t>::insert(const DVec & pos,const val_t & value)
+{
+	insert(CoverTreeNode<val_t>(pos,value));
+}
+
+template <typename val_t>
+void CTHist<val_t>::insert(const CoverTreeNode<val_t> & x)
+{
+	if (_nodes.size() == 0)
+	{
+		for (unsigned int i = 0; i < x.pos.size(); i++)
+		{
+			_lower_limits[i] = x.pos[i];
+			_upper_limits[i] = x.pos[i];
+		}
+	}
+	else
+	{
+		for (unsigned int i = 0; i < x.pos.size(); i++)
+		{
+			if (x.pos[i] < _lower_limits[i])
+				_lower_limits[i] = x.pos[i];
+			if (x.pos[i] > _upper_limits[i])
+				_upper_limits[i] = x.pos[i];
+		}
+	}
+
+	if (_nodes.size() == _max_size)
+		insertMerge(x);
+	else
+		CoverTree<val_t>::insert(x);
+}
+
+template <typename val_t>
+double CTHist<val_t>::epsilon = 0.001;
+
+template <typename val_t>
+void CTHist<val_t>::insertMerge(const CoverTreeNode<val_t> & x)
+{
+	int level;
+	int parent = -1;
+	int nearest_idx = findNearestNeighbor_(x,_root_idx,_root_idx,
+										   _root_level,level,parent);
+	CoverTreeNode<val_t> & nearest = _nodes[nearest_idx];
+	if (dist(nearest,x) <= epsilon)
+		nearest.value += x.value;
+	else
+	{
+		nearest = mergeNode(x,nearest);
+
+		std::vector<int> desc = std::vector<int>();
+		descendants(nearest,desc);
+
+		//If we aren't the root and we moved outside the coverdistance
+		if (parent != -1 && dist(nearest,_nodes[parent]) > covdist(level+1))
+		{
+			//Remove our node from the parent
+			_nodes[parent].children.erase(find(_nodes[parent].children.begin(),_nodes[parent].children.end(),nearest_idx));
+
+			//Add all descendants to the parent
+			for (int c_idx : desc)
+			{
+				_nodes[c_idx].children.clear();
+				insert_rec(c_idx,_nodes[parent],level+1);
+			}
+			_nodes[nearest_idx].children.clear();
+
+			//Readd our node
```

Comment:
It is conventional, in C++, to have a single blank space after the `//` delimiter.